replica/table.cc

```diff
@@ -2053,9 +2063,13 @@ future<> tablet_storage_group_manager::handle_tablet_split_completion(const loca
             table_id, new_tablet_count, old_tablet_count * split_size));
     }
 
+    // Even though the loop over storage groups does not erase map's elements, it stops their components
+    // and so we need to hold exclusive lock throughout the whole process.
+    rwlock::holder exclusive_lock = co_await get_rwlock().hold_write_lock();
     // Stop the released main compaction groups asynchronously
     future<> stop_fut = make_ready_future<>();
-    for (auto& [id, sg] : storage_groups()) {
+    // Do not use for_each_storage_group to prevent preemption points.
```

Comment:
If we're holding the exclusive_lock above why does this matter?

Reply:
I have the same question.

Reply:
Writing that I had in mind a situation where:
The loop stops some cgs, preempts, and then some other layer accesses single storage group (or cg) and meets the stopped one.

But here we actually rewrite storage_group map, so this definitely shouldn't happen (otherwise we would hang around with some dangling groups).